rust/sbd-server/src/lib.rs

```diff
@@ -233,11 +233,78 @@ impl SbdServer {
         // read this as a prioritization of existing connections over incoming
         let connect_limit = Arc::new(tokio::sync::Semaphore::new(1024));
 
-        let weak_cslot = cslot.weak();
+        let mut bind_port_zero = Vec::new();
+        let mut bind_explicit_port = Vec::new();
+
         for bind in config.bind.iter() {
             let a: std::net::SocketAddr = bind.parse().map_err(Error::other)?;
 
-            let tcp = tokio::net::TcpListener::bind(a).await?;
+            if a.port() == 0 {
+                bind_port_zero.push(a);
+            } else {
+                bind_explicit_port.push(a);
+            }
+        }
+
+        let (mut listeners, mut l2) = tokio::join!(
+            async {
+                // bail if there are no zero port bindings
+                if bind_port_zero.is_empty() {
+                    return Vec::new();
+                }
+
+                // try twice to re-use port
+                'top: for _ in 0..2 {
+                    let mut listeners = Vec::new();
+
+                    let mut a_iter = bind_port_zero.iter();
+
+                    let a = a_iter.next().unwrap();
+                    if let Ok(tcp) = tokio::net::TcpListener::bind(a).await {
+                        let port = tcp.local_addr().unwrap().port();
+                        listeners.push(tcp);
+
+                        for a in a_iter {
+                            let mut a = *a;
+                            a.set_port(port);
+                            match tokio::net::TcpListener::bind(a).await {
+                                Err(_) => continue 'top,
+                                Ok(tcp) => listeners.push(tcp),
+                            }
+                        }
+
+                        return listeners;
```

Comment:
Shouldn't you forget the exit condition here and just rebind the ports you have already bound ;)